tests/trainer/test_trainer.py

```diff
@@ -1594,6 +1595,90 @@ def test_fp16_full_eval(self):
         # perfect world: fp32_init/2 == fp16_eval
         self.assertAlmostEqual(fp16_eval, fp32_init / 2, delta=5_000)
 
+    @require_torch_gpu
+    @require_torchdynamo
+    def test_torchdynamo_full_eval(self):
+        n_gpus = get_gpu_count()
+
+        bs = 8
+        eval_len = 16 * n_gpus
+        # make the params somewhat big so that there will be enough RAM consumed to be able to
+        # measure things. We should get about 64KB for a+b in fp32
+        a = torch.ones(1000, bs) + 0.001
+        b = torch.ones(1000, bs) - 0.001
+
+        # 1. Default - without TorchDynamo
+        trainer = get_regression_trainer(a=a, b=b, eval_len=eval_len)
+        metrics = trainer.evaluate()
+        original_eval_loss = metrics["eval_loss"]
+        del trainer
+
+        # 2. TorchDynamo eager
+        trainer = get_regression_trainer(a=a, b=b, eval_len=eval_len, torchdynamo="eager")
+        metrics = trainer.evaluate()
+        self.assertAlmostEqual(metrics["eval_loss"], original_eval_loss)
+        del trainer
+
+        # 3. TorchDynamo nvfuser
+        trainer = get_regression_trainer(a=a, b=b, eval_len=eval_len, torchdynamo="nvfuser")
+        metrics = trainer.evaluate()
+        self.assertAlmostEqual(metrics["eval_loss"], original_eval_loss)
+
+    @require_torch_gpu
+    @require_torchdynamo
+    def test_torchdynamo_memory(self):
+        class MyModule(torch.nn.Module):
+            """Simple module that does aggressive fusion"""
+
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x):
+                for _ in range(20):
```

Comment:
@stas00 Can you try changing this number to 10 to see if the test improves? If it doesn't, lets make it 1 to see if this is the culprit. At 1, the test will fail, but we will have little more info to debug,

Reply:
- 1 or 2 fails as expected
- 5 succeeds
- 10 hangs - same stack

Reply:
I figured it out - Trainer tried to run 2 gpus DP and somehow was hanging.

Forcing one gpu breaks past this hanging issue

```
CUDA_VISIBLE_DEVICES=0 pyt tests/trainer/test_trainer.py -k torchdynamo_memory -sv
```
now fails:
```
self = <tests.trainer.test_trainer.TrainerIntegrationTest testMethod=test_torchdynamo_memory>

    @require_torch_gpu
    @require_torchdynamo
    def test_torchdynamo_memory(self):
        class MyModule(torch.nn.Module):
            """Simple module that does aggressive fusion"""
    
            def __init__(self):
                super().__init__()
    
            def forward(self, x):
                for _ in range(20):
                    x = torch.nn.functional.relu(x)
                return x
    
        mod = MyModule()
    
        # 1. Default - without TorchDynamo
        a = torch.ones(1024, 1024, device="cuda", requires_grad=True)
        a.grad = None
        trainer = Trainer(model=mod)
        # warmup
        for _ in range(10):
>           orig_loss = trainer.training_step(mod, {"x": a})

tests/trainer/test_trainer.py:1649: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/trainer.py:2263: in training_step
    loss.backward()
/home/stas/anaconda3/envs/py38-pt112/lib/python3.8/site-packages/torch/_tensor.py:399: in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
/home/stas/anaconda3/envs/py38-pt112/lib/python3.8/site-packages/torch/autograd/__init__.py:166: in backward
    grad_tensors_ = _make_grads(tensors, grad_tensors_, is_grads_batched=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

outputs = (tensor([1., 1., 1.,  ..., 1., 1., 1.], device='cuda:0',
       grad_fn=<SelectBackward0>),), grads = (None,)
is_grads_batched = False

    def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor],
                    is_grads_batched: bool) -> Tuple[_OptionalTensor, ...]:
        new_grads: List[_OptionalTensor] = []
        for out, grad in zip(outputs, grads):
            if isinstance(grad, torch.Tensor):
                grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
                if not out.shape == grad_shape:
                    if is_grads_batched:
                        raise RuntimeError("If `is_grads_batched=True`, we interpret the first "
                                           "dimension of each grad_output as the batch dimension. "
                                           "The sizes of the remaining dimensions are expected to match "
                                           "the shape of corresponding output, but a mismatch "
                                           "was detected: grad_output["
                                           + str(grads.index(grad)) + "] has a shape of "
                                           + str(grad.shape) + " and output["
                                           + str(outputs.index(out)) + "] has a shape of "
                                           + str(out.shape) + ". "
                                           "If you only want some tensors in `grad_output` to be considered "
                                           "batched, consider using vmap.")
                    else:
                        raise RuntimeError("Mismatch in shape: grad_output["
                                           + str(grads.index(grad)) + "] has a shape of "
                                           + str(grad.shape) + " and output["
                                           + str(outputs.index(out)) + "] has a shape of "
                                           + str(out.shape) + ".")
                if out.dtype.is_complex != grad.dtype.is_complex:
                    raise RuntimeError("For complex Tensors, both grad_output and output"
                                       " are required to have the same dtype."
                                       " Mismatch in dtype: grad_output["
                                       + str(grads.index(grad)) + "] has a dtype of "
                                       + str(grad.dtype) + " and output["
                                       + str(outputs.index(out)) + "] has a dtype of "
                                       + str(out.dtype) + ".")
                new_grads.append(grad)
            elif grad is None:
                if out.requires_grad:
                    if out.numel() != 1:
>                       raise RuntimeError("grad can be implicitly created only for scalar outputs")
E                       RuntimeError: grad can be implicitly created only for scalar outputs

/home/stas/anaconda3/envs/py38-pt112/lib/python3.8/site-packages/torch/autograd/__init__.py:67: RuntimeError
```

Reply:
I see. For single GPU, the `training_step` was not calling `mean()` on the output. PyTorch expects a scalar loss for `.backward()` call and therefore you saw the error message you pasted. I just added a CustomTrainer to reduce to scalar and the test passes for single GPU as well.

I am not sure why it hangs for DP though. Is it possible to limit the usage only in the case of single GPUs? We have not really tested TorchDynamo for distributed training.

Reply:
I pushed a change in decorators that will now require a single gpu. But you probably want to adapt the first test not to rely on `get_gpu_count()` then as it'll always be 1.

Also it is probably a good idea to add to the arguments entry that DP/DDP is not supported as of this moment?

Reply:
ok, I had to fetch the latest pytorch nightly - apparently the 1 week old one wasn't cutting it.

now it all works.

Reply:
it works on 2 gpus with DP as well - should we remove the restrictions then and put it back how you coded it originally?

Reply:
The hanging was on a single GPU as well with pytorch from 05-18.